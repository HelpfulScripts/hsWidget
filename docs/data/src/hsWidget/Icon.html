
                <html><style>body { overflow:hidden;}
            h1 { font-family: Arial, sans-serif; font-size: 24px; color: #44a; }
            p { margin:0; padding:0; }
            br  { margin:0; padding:0; }
            .line { 
                white-space: nowrap;
                height:16px; 
            }
            .line>span { 
                display:inline-block; background-color:#eef; height:100%; 
                margin: 0 5px 0 0; padding-right: 5px; color:#999;   
            }
            comment {color: #080;} module {color: #804;} 
            quote {color: #008;} comment>quote {color: #080;}
            .listing { margin: 10px; border: 1px solid #ccc; 
                    font-family: SFMono-Regular, Consolas, 'Liberation Mono', Menlo, Courier, monospace;
                    font-size: 14px;  
                    overflow:scroll;
                    height:90%;
            }
            code { padding: 5px 0;}</style></html><body>
                <h1>Icon.ts</h1>
                <div class='listing'><code><p id=1 class="line"><span>&nbsp;&nbsp;&nbsp;1</span><comment>/**</comment></p>
<p id=2 class="line"><span>&nbsp;&nbsp;&nbsp;2</span><comment>&nbsp;*&nbsp;#&nbsp;Icon&nbsp;Widget</comment></p>
<p id=3 class="line"><span>&nbsp;&nbsp;&nbsp;3</span><comment>&nbsp;*&nbsp;Shows&nbsp;an&nbsp;icon&nbsp;with&nbsp;optional&nbsp;text&nbsp;label</comment></p>
<p id=4 class="line"><span>&nbsp;&nbsp;&nbsp;4</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=5 class="line"><span>&nbsp;&nbsp;&nbsp;5</span><comment>&nbsp;*&nbsp;###&nbsp;Profile</comment></p>
<p id=6 class="line"><span>&nbsp;&nbsp;&nbsp;6</span><comment>&nbsp;*&nbsp;invoked&nbsp;as&nbsp;`m(Icon,&nbsp;{mdi:<string>&nbsp;,&nbsp;text:<string>});`</comment></p>
<p id=7 class="line"><span>&nbsp;&nbsp;&nbsp;7</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=8 class="line"><span>&nbsp;&nbsp;&nbsp;8</span><comment>&nbsp;*&nbsp;###&nbsp;Attributes&nbsp;(node.attrs):</comment></p>
<p id=9 class="line"><span>&nbsp;&nbsp;&nbsp;9</span><comment>&nbsp;*&nbsp;-&nbsp;`text:&nbsp;string`&nbsp;name&nbsp;to&nbsp;show&nbsp;as&nbsp;button&nbsp;text</comment></p>
<p id=10 class="line"><span>&nbsp;&nbsp;10</span><comment>&nbsp;*&nbsp;-&nbsp;`mdi:&nbsp;string`&nbsp;MDI&nbsp;icon&nbsp;name&nbsp;(see:&nbsp;https:<comment>//materialdesignicons.com/)</comment></comment></p>
<p id=11 class="line"><span>&nbsp;&nbsp;11</span><comment><comment></comment>&nbsp;*&nbsp;-&nbsp;`css:&nbsp;string`&nbsp;css&nbsp;class&nbsp;to&nbsp;assign&nbsp;to&nbsp;button&nbsp;tag</comment></p>
<p id=12 class="line"><span>&nbsp;&nbsp;12</span><comment>&nbsp;*&nbsp;-&nbsp;`style:&nbsp;string`&nbsp;style&nbsp;string&nbsp;to&nbsp;apply&nbsp;to&nbsp;button&nbsp;tag</comment></p>
<p id=13 class="line"><span>&nbsp;&nbsp;13</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=14 class="line"><span>&nbsp;&nbsp;14</span><comment>&nbsp;*&nbsp;###&nbsp;Example</comment></p>
<p id=15 class="line"><span>&nbsp;&nbsp;15</span><comment>&nbsp;*&nbsp;<example></comment></p>
<p id=16 class="line"><span>&nbsp;&nbsp;16</span><comment>&nbsp;*&nbsp;<file&nbsp;name=<quote>'script.js'</quote>>&nbsp;</comment></p>
<p id=17 class="line"><span>&nbsp;&nbsp;17</span><comment>&nbsp;*&nbsp;m.mount(root,&nbsp;{view:&nbsp;()&nbsp;=>&nbsp;m(hsWidget.Label,&nbsp;{</comment></p>
<p id=18 class="line"><span>&nbsp;&nbsp;18</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;css:&nbsp;<quote>'.myLabel'</quote>,&nbsp;</comment></p>
<p id=19 class="line"><span>&nbsp;&nbsp;19</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;style:&nbsp;<quote>'text-align:&nbsp;right;'</quote>,</comment></p>
<p id=20 class="line"><span>&nbsp;&nbsp;20</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;text:&nbsp;<quote>'This&nbsp;is&nbsp;a&nbsp;<b>label</b>'</quote>,</comment></p>
<p id=21 class="line"><span>&nbsp;&nbsp;21</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;mdi:&nbsp;<quote>'leaf'</quote></comment></p>
<p id=22 class="line"><span>&nbsp;&nbsp;22</span><comment>&nbsp;*&nbsp;})});</comment></p>
<p id=23 class="line"><span>&nbsp;&nbsp;23</span><comment>&nbsp;*&nbsp;</file></comment></p>
<p id=24 class="line"><span>&nbsp;&nbsp;24</span><comment>&nbsp;*&nbsp;<file&nbsp;name=<quote>'style.css'</quote>></comment></p>
<p id=25 class="line"><span>&nbsp;&nbsp;25</span><comment>&nbsp;*&nbsp;.myLabel&nbsp;{&nbsp;&nbsp;</comment></p>
<p id=26 class="line"><span>&nbsp;&nbsp;26</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;background-color:&nbsp;white;&nbsp;</comment></p>
<p id=27 class="line"><span>&nbsp;&nbsp;27</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;margin:&nbsp;5px;</comment></p>
<p id=28 class="line"><span>&nbsp;&nbsp;28</span><comment>&nbsp;*&nbsp;}</comment></p>
<p id=29 class="line"><span>&nbsp;&nbsp;29</span><comment>&nbsp;*&nbsp;</file></comment></p>
<p id=30 class="line"><span>&nbsp;&nbsp;30</span><comment>&nbsp;*&nbsp;</example></comment></p>
<p id=31 class="line"><span>&nbsp;&nbsp;31</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=32 class="line"><span>&nbsp;&nbsp;32</span><comment>&nbsp;*/</comment></p>
<p id=33 class="line"><span>&nbsp;&nbsp;33</span></p>
<p id=34 class="line"><span>&nbsp;&nbsp;34</span><comment>/**&nbsp;*/</comment></p>
<p id=35 class="line"><span>&nbsp;&nbsp;35</span>import&nbsp;{&nbsp;m,&nbsp;Vnode&nbsp;}&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;from&nbsp;<quote>'hslayout'</quote>;</p>
<p id=36 class="line"><span>&nbsp;&nbsp;36</span></p>
<p id=37 class="line"><span>&nbsp;&nbsp;37</span><comment>/**</comment></p>
<p id=38 class="line"><span>&nbsp;&nbsp;38</span><comment>&nbsp;*&nbsp;#&nbsp;Icon&nbsp;</comment></p>
<p id=39 class="line"><span>&nbsp;&nbsp;39</span><comment>&nbsp;*&nbsp;An&nbsp;MDI&nbsp;icon&nbsp;with&nbsp;text</comment></p>
<p id=40 class="line"><span>&nbsp;&nbsp;40</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=41 class="line"><span>&nbsp;&nbsp;41</span><comment>&nbsp;*&nbsp;###&nbsp;Profile</comment></p>
<p id=42 class="line"><span>&nbsp;&nbsp;42</span><comment>&nbsp;*&nbsp;invoked&nbsp;as&nbsp;`m(Icon,&nbsp;{mdi:<string>,&nbsp;text:<string>});`</comment></p>
<p id=43 class="line"><span>&nbsp;&nbsp;43</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=44 class="line"><span>&nbsp;&nbsp;44</span><comment>&nbsp;*&nbsp;###&nbsp;Attributes&nbsp;(node.attrs):</comment></p>
<p id=45 class="line"><span>&nbsp;&nbsp;45</span><comment>&nbsp;*&nbsp;-&nbsp;`text?:&nbsp;string`&nbsp;the&nbsp;label&nbsp;text;&nbsp;</comment></p>
<p id=46 class="line"><span>&nbsp;&nbsp;46</span><comment>&nbsp;*/</comment></p>
<p id=47 class="line"><span>&nbsp;&nbsp;47</span>export&nbsp;class&nbsp;Icon&nbsp;{</p>
<p id=48 class="line"><span>&nbsp;&nbsp;48</span>&nbsp;&nbsp;&nbsp;&nbsp;view(node:&nbsp;Vnode):&nbsp;Vnode&nbsp;{&nbsp;</p>
<p id=49 class="line"><span>&nbsp;&nbsp;49</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;const&nbsp;css&nbsp;=&nbsp;node.attrs.css&nbsp;||&nbsp;<quote>''</quote>;</p>
<p id=50 class="line"><span>&nbsp;&nbsp;50</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;const&nbsp;style&nbsp;=&nbsp;node.attrs.style&nbsp;||&nbsp;<quote>''</quote>;</p>
<p id=51 class="line"><span>&nbsp;&nbsp;51</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;const&nbsp;text&nbsp;&nbsp;=&nbsp;node.attrs.text&nbsp;||&nbsp;<quote>'unspecified'</quote>;</p>
<p id=52 class="line"><span>&nbsp;&nbsp;52</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;const&nbsp;mdi&nbsp;&nbsp;&nbsp;=&nbsp;node.attrs.mdi;</p>
<p id=53 class="line"><span>&nbsp;&nbsp;53</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return&nbsp;m(`.hs-label&nbsp;${css}`,&nbsp;{&nbsp;style:style&nbsp;},&nbsp;m.trust(text));&nbsp;</p>
<p id=54 class="line"><span>&nbsp;&nbsp;54</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=55 class="line"><span>&nbsp;&nbsp;55</span>}</p>
<p id=56 class="line"><span>&nbsp;&nbsp;56</span></p>
<p id=57 class="line"><span>&nbsp;&nbsp;57</span></p></code></div>
                </body>
            