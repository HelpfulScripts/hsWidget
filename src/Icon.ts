/**
 * # Icon widget
 * Shows an [MDI](https://materialdesignicons.com/) icon with optional text label.
 * The icon will scale to fill the height of the element, which defaults to `1em`.
 * The content will vertically align centered if the icon is higher than the text.
 * 
 * If a predefined icon is used, the provided `mdi` key will als be used as a css class
 * selector for the icon. This allows for a semantic coloring of the icon via CSS.
 * 
 * ### Profile
 * invoked as `m(Icon, <IconAttrs>, <content>);`.
 * 
 * See: {@link Icon.IconAttrs IconAttrs}
 * 
 * ### Example
 * <example>
 * <file name='script.js'>ximport='Icon.x.js'</file>
 * <file name='style.css'>
 * .hs-execution .hs_grid.hs_grid_row { height:100%; }
 * .hs-execution .hs_grid.hs_grid_col { width:100%; }
 * </file>
 * </example>
 * 
 * ### Predefined icons:
 * - 'info'
 * - 'warn'
 * - 'stop'
 */

/** */
import { WidgetAttrs }      from './Widget';
import { Widget }           from './Widget';
import m                    from "mithril";
import { Vnode }            from "./Widget";
import { mdiAlert }         from '@mdi/js';
import { mdiInformation }   from '@mdi/js';
import { mdiAlertOctagon }  from '@mdi/js';
import { mdiPower }         from '@mdi/js';
import { mdiPlus }          from '@mdi/js';
import { mdiMinus }         from '@mdi/js';
import { mdiChevronUp }     from '@mdi/js';
import { mdiChevronDown }   from '@mdi/js';
import { mdiChevronLeft }   from '@mdi/js';
import { mdiChevronRight }  from '@mdi/js';
import { mdiHomeOutline }   from '@mdi/js';

export interface IconAttrs extends WidgetAttrs {
    /** 
     * optional [MDI](https://materialdesignicons.com/) icon specifier, defaults to 'info'
     * - either an `svg` string for the icon, as imported from `@mdi/js`;
     * - or a key for one of the {@link Icon Predefined Icons};
     */
    mdi?: string;
}


/**
 * # Icon 
 * An MDI icon with text, invoked as 
 * `m(Icon, <{@link Icon.IconAttrs `IconAttrs`}>, <content>);`
 */
export class Icon extends Widget {
    static predefined = {
        info:           mdiInformation,
        warn:           mdiAlert,
        stop:           mdiAlertOctagon,
        power:          mdiPower,
        plus:           mdiPlus,
        minus:          mdiMinus,
        chevronUp:      mdiChevronUp,
        chevronDown:    mdiChevronDown,
        chevronLeft:    mdiChevronLeft,
        chevronRight:   mdiChevronRight,
        home:           mdiHomeOutline
    };
    view(node: Vnode<IconAttrs, this>): m.Children { 
        const svg = node.attrs.mdi || 'info';
        const icon = Icon.predefined[svg] || svg;
        // set the cvg css if `mdi` has a predefined key
        const css = Icon.predefined[svg]? `.${svg}` : '';
        return m(`.hs_icon`, this.attrs(node.attrs), [
            m(`svg${css}`, { height:`100%`, viewBox:`0 0 24 24`}, 
                m('path', {d:icon})),
            // avoid .hs_icon_content formatting if no children:
            (<any[]>node.children).length? m('span.hs_icon_content', node.children) : undefined
        ]);
    }
}
