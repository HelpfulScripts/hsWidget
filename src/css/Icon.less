@iconStrokeColor: #443;

.hs_icon {
    height: 1.5em;
    fill: @iconStrokeColor;
    stroke: @iconStrokeColor;
    .info { fill: blue; stroke-width: 0;}
    .warn { fill: yellow; stroke:#cccc00; stroke-width: 0;}
    .stop { fill: red; stroke-width: 0;}
    .hs_icon_content {
        padding-left: 5px;
        height:100%;
        line-height:100%;
        position:relative;
        bottom: 50%;
        display: inline-block;
        translate: 0 0.35em;
        vertical-align: baseline;
    }
}