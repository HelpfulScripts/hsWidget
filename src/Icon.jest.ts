import m                from "mithril";
import { Icon }         from './Icon';


const root = window.document.createElement("div");
describe('Icon', () => {
    beforeAll(()=>{
        m.mount(root, {view: () => 
            m('.white', {style:'background-color: white;'}, 
                Object.keys(Icon.predefined).map(k => m(Icon, { mdi:k}, k))
            )
        });
    });
    it('should match snapshot', () => {
        expect(root).toMatchSnapshot();
    });
});
   