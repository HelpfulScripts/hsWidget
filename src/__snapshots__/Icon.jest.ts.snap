// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon should match snapshot 1`] = `
<div>
  <div
    class="white"
    style="background-color: white;"
  >
    <div
      class="hs_icon"
    >
      <svg
        class="info"
        height="100%"
        viewBox="0 0 24 24"
      >
        <path
          d="M13,9H11V7H13M13,17H11V11H13M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2Z"
        />
      </svg>
      <span
        class="hs_icon_content"
      >
        info
      </span>
    </div>
    <div
      class="hs_icon"
    >
      <svg
        class="warn"
        height="100%"
        viewBox="0 0 24 24"
      >
        <path
          d="M13 14H11V9H13M13 18H11V16H13M1 21H23L12 2L1 21Z"
        />
      </svg>
      <span
        class="hs_icon_content"
      >
        warn
      </span>
    </div>
    <div
      class="hs_icon"
    >
      <svg
        class="stop"
        height="100%"
        viewBox="0 0 24 24"
      >
        <path
          d="M13 13H11V7H13M11 15H13V17H11M15.73 3H8.27L3 8.27V15.73L8.27 21H15.73L21 15.73V8.27L15.73 3Z"
        />
      </svg>
      <span
        class="hs_icon_content"
      >
        stop
      </span>
    </div>
    <div
      class="hs_icon"
    >
      <svg
        class="power"
        height="100%"
        viewBox="0 0 24 24"
      >
        <path
          d="M16.56,5.44L15.11,6.89C16.84,7.94 18,9.83 18,12A6,6 0 0,1 12,18A6,6 0 0,1 6,12C6,9.83 7.16,7.94 8.88,6.88L7.44,5.44C5.36,6.88 4,9.28 4,12A8,8 0 0,0 12,20A8,8 0 0,0 20,12C20,9.28 18.64,6.88 16.56,5.44M13,3H11V13H13"
        />
      </svg>
      <span
        class="hs_icon_content"
      >
        power
      </span>
    </div>
    <div
      class="hs_icon"
    >
      <svg
        class="plus"
        height="100%"
        viewBox="0 0 24 24"
      >
        <path
          d="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
        />
      </svg>
      <span
        class="hs_icon_content"
      >
        plus
      </span>
    </div>
    <div
      class="hs_icon"
    >
      <svg
        class="minus"
        height="100%"
        viewBox="0 0 24 24"
      >
        <path
          d="M19,13H5V11H19V13Z"
        />
      </svg>
      <span
        class="hs_icon_content"
      >
        minus
      </span>
    </div>
    <div
      class="hs_icon"
    >
      <svg
        class="plusminus"
        height="100%"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 7H6V4H8V7H11V9H8V12H6V9H3V7M13 15H21V17H13V15M16.04 3H18.35L7.96 21H5.65L16.04 3Z"
        />
      </svg>
      <span
        class="hs_icon_content"
      >
        plusminus
      </span>
    </div>
    <div
      class="hs_icon"
    >
      <svg
        class="chevronUp"
        height="100%"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.41,15.41L12,10.83L16.59,15.41L18,14L12,8L6,14L7.41,15.41Z"
        />
      </svg>
      <span
        class="hs_icon_content"
      >
        chevronUp
      </span>
    </div>
    <div
      class="hs_icon"
    >
      <svg
        class="chevronDown"
        height="100%"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
        />
      </svg>
      <span
        class="hs_icon_content"
      >
        chevronDown
      </span>
    </div>
    <div
      class="hs_icon"
    >
      <svg
        class="chevronLeft"
        height="100%"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.41,16.58L10.83,12L15.41,7.41L14,6L8,12L14,18L15.41,16.58Z"
        />
      </svg>
      <span
        class="hs_icon_content"
      >
        chevronLeft
      </span>
    </div>
    <div
      class="hs_icon"
    >
      <svg
        class="chevronRight"
        height="100%"
        viewBox="0 0 24 24"
      >
        <path
          d="M8.59,16.58L13.17,12L8.59,7.41L10,6L16,12L10,18L8.59,16.58Z"
        />
      </svg>
      <span
        class="hs_icon_content"
      >
        chevronRight
      </span>
    </div>
    <div
      class="hs_icon"
    >
      <svg
        class="home"
        height="100%"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 5.69L17 10.19V18H15V12H9V18H7V10.19L12 5.69M12 3L2 12H5V20H11V14H13V20H19V12H22L12 3Z"
        />
      </svg>
      <span
        class="hs_icon_content"
      >
        home
      </span>
    </div>
  </div>
</div>
`;
